perllib/FixMyStreet/DB/ResultSet/Problem.pm

```diff
@@ -15,15 +15,41 @@ sub set_restriction {
     $site_key = $key;
 }
 
-sub to_body {
-    my ($rs, $bodies, $join) = @_;
-    return $rs unless $bodies;
+sub body_query {
+    my ($rs, $bodies) = @_;
     unless (ref $bodies eq 'ARRAY') {
         $bodies = [ map { ref $_ ? $_->id : $_ } $bodies ];
     }
+    \[ "regexp_split_to_array(bodies_str, ',') && ?", [ {} => $bodies ] ]
+}
+
+# Edits PARAMS in place to either hide non_public reports, or show them
+# if user is superuser (all) or inspector (correct body)
+sub non_public_if_possible {
+    my ($rs, $params, $c) = @_;
+    if ($c->user_exists) {
+        if ($c->user->is_superuser) {
+            # See all reports, no restriction
+        } elsif ($c->user->has_body_permission_to('report_inspect')) {
+            $params->{'-or'} = [
+                non_public => 0,
+                $rs->body_query($c->user->from_body->id),
+            ];
+        } else {
+            $params->{non_public} = 0;
+        }
+    } else {
+        $params->{non_public} = 0;
+    }
+}
+
+sub to_body {
+    my ($rs, $bodies, $join) = @_;
+    return $rs unless $bodies;
     $join = { join => 'problem' } if $join;
+    # This isn't using $rs here because $rs might be Problem, Comment, or Nearby
```

Comment:
I'm assuming this means it's not using `$rs->body_query`? Could you make that clearer in the comment, as `$rs` does appear on the next line and it took me a minute.